### PR TITLE
Fix time-val input buttons to change calendar date-range

### LINKED ISSUE #25966: [BO] Dashboard time-val selection doesn't change the calendar
Hi there,

Selecting a time-val (day, month, year,...) from the calendar form, doesn't change the date-picker and causes a jquery error.

1. Go to BO > Dashboard
2. Select a time-val from calendar form 

image

3. See error in browser console

**Screenshots**

Screenshot 2021-09-22 at 12-11-40 پیشخوان • ONE

Screenshot 2021-09-22 121231

#### Additional information

* PrestaShop version: 1.7.8.0 (RTL)
* PHP version: 7.4

### PR DESCRIPTION
**Pull request by mparvazi**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 1.7.8.x
| Description?      | Selecting a time-val button (day, month, year,...), doesn't change the calendar date-range and causes a jquery error.
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #25966
| How to test?      | Select a time-val button to change the calendar date-range.
| Possible impacts? | -

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/26145)

### CODE REVIEW

**Comment by mparvazi** [comment]
A redo for PR #25986

**Review by matks — DISMISSED** [review]

**Review by matthieu-rolland — COMMENTED** [review]
hi @mparvazi , thank you for your contribution 👍 

It seems the reason removing `const` fixes the bug is because it's block-scoped, and those variables are then used in event callbacks. 

But if I'm not wrong, it's a good practice to properly declare the scope of variables, maybe @NeOMakinG can confirm? I could be wrong...

Declaring those 2 variables as `var datepickerStart` or `window.datepickerStart` should fix your bug as well.

**Comment by mparvazi** [comment]
Hi @matthieu-rolland
It works fine with `window.datepickerStart` but `var datepickerStart` throws an error as like as `const datepickerStart`.
I tested by `let datepickerStart` too, but same error occurs.

I think `datepickerStart` and `datepickerEnd` should be declare outside of  `$(document).ready()` and initiate inside it.
These variables are used in other functions outside of `$(document).ready()`. That's why removing `const` or using `window.datepickerStart` makes the script run correctly, because in this case they are considered as global variables.

**Comment by NeOMakinG** [comment]
@mparvazi what's the exact error output?

**Review by Progi1984 — DISMISSED** [review]

**Comment by mparvazi** [comment]
@NeOMakinG 

dashboard

console

code

change

Putting their definition outside  or simply remove `const` will fix the problem, so they are considered as global variables and can use in other functions (e.g. `updatePickerFromInput()`)

**Comment by NeOMakinG** [comment]
Oh yeah, just declare these variable outside of the `document.ready`, it looks fine to me honestly

File: admin-dev/themes/default/js/calendar.js [hunk 1]
```diff
@@ -307,9 +307,12 @@ function setPreviousYear() {
   $('#date-end-compare').val(endDate.format($('#date-start').data('date-format')));
 }
 
+var datepickerStart;
+var datepickerEnd;
```
Comment (NeOMakinG) [inline admin-dev/themes/default/js/calendar.js]:
```suggestion
let datepickerStart;
let datepickerEnd;
```

When these vars are out of the scope, do it works with let? I think so

**Comment by mparvazi** [comment]
@NeOMakinG 
Works fine with `let`

**Review by NeOMakinG — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @mparvazi,

I have an issue: in the Dashboard Date, compare period can't be selected automatically.
image
I attached a screen record
https://watch.screencastify.com/v/exxUARgbaHbUppsuomg5

Previously fixed here: https://github.com/PrestaShop/PrestaShop/pull/22997/files

Thanks!

**Comment by mparvazi** [comment]
Hello @khouloudbelguith 

The commit at PR #22997 already exists in `calendar.js`

The problem is related to `===` operator, it compares both the value and the type:
`console.log(2 === '2');` => false
The operator `==` checks equality only after converting both values to a common type:
`console.log(2 == '2');` => true

My suggestion:
 Line:405 `if (this.value === 1) setPreviousPeriod();` => `if (this.value == 1) setPreviousPeriod();`
Line:407 `if (this.value === 2) setPreviousYear();` => `if (this.value == 2) setPreviousYear();` 
Line:416 `if (this.value === 3) $('#date-start-compare').focus();` => `if (this.value == 3) $('#date-start-compare').focus();`

I tested these changes in `calendar.js` and the problem is solved.
https://watch.screencastify.com/v/K7z4SLaoHrZPRyY9QM91

**Comment by Progi1984** [comment]
Ping @NeOMakinG for helping @mparvazi

**Comment by NeOMakinG** [comment]
@mparvazi I would suggest to compare with a string instead of ditching the type compare, something like `this.value === '1'` as an input value is always a string anyway

**Review by NeOMakinG — APPROVED** [review]

**Review by atomiix — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hi @mparvazi,

It is ok :heavy_check_mark: 
https://watch.screencastify.com/v/avAYUmcoJgNVXXTUnf7O

Thanks!

**Comment by Progi1984** [comment]
Thanks @mparvazi & @khouloudbelguith

**Comment by lacell75** [comment]
I reproduced this issue in v1.7.8.1 yet
image

**Comment by mparvazi** [comment]
@lacell75 Please clear the browser cache.

**Comment by khouloudbelguith** [comment]
Hi @lacell75,

I tried with PS1781 fresh install => ok
https://watch.screencastify.com/v/Rp8QvmicNdwWbTt9a30y

Thanks!

**Comment by lacell75** [comment]
Thanks! that works after cash refresh

Le lun. 22 nov. 2021 à 21:28, M. Parvazi ***@***.***> a
écrit :

**Pull request merged**